urlcdab7d8364_jrwalsh_CycTools_tgz-pJ8-edu_iastate_cyctools_MainJ8/src/edu/iastate/cyctools/externalSourceCode/DiffMatchPatch.java

```diff
@@ -0,0 +1,2535 @@
+              + nextDiff.text;
+          pointer.remove(); // Delete nextDiff.
+          nextDiff = pointer.hasNext() ? pointer.next() : null;
+          changes = true;
+        }
+      }
+      prevDiff = thisDiff;
+      thisDiff = nextDiff;
+      nextDiff = pointer.hasNext() ? pointer.next() : null;
+    }
+    // If shifts were made, the diff needs reordering and another shift sweep.
+    if (changes) {
+      diff_cleanupMerge(diffs);
+    }
+  }
+
+  /**
+   * loc is a location in text1, compute and return the equivalent location in
+   * text2.
+   * e.g. "The cat" vs "The big cat", 1->1, 5->8
+   * @param diffs LinkedList of Diff objects.
+   * @param loc Location within text1.
+   * @return Location within text2.
+   */
+  public int diff_xIndex(LinkedList<Diff> diffs, int loc) {
+    int chars1 = 0;
+    int chars2 = 0;
+    int last_chars1 = 0;
+    int last_chars2 = 0;
+    Diff lastDiff = null;
+    for (Diff aDiff : diffs) {
+      if (aDiff.operation != Operation.INSERT) {
+        // Equality or deletion.
+        chars1 += aDiff.text.length();
+      }
+      if (aDiff.operation != Operation.DELETE) {
+        // Equality or insertion.
+        chars2 += aDiff.text.length();
+      }
+      if (chars1 > loc) {
+        // Overshot the location.
+        lastDiff = aDiff;
+        break;
+      }
+      last_chars1 = chars1;
+      last_chars2 = chars2;
+    }
+    if (lastDiff != null && lastDiff.operation == Operation.DELETE) {
+      // The location was deleted.
+      return last_chars2;
+    }
+    // Add the remaining character length.
+    return last_chars2 + (loc - last_chars1);
+  }
+
+  /**
+   * Convert a Diff list into a pretty HTML report.
+   * @param diffs LinkedList of Diff objects.
+   * @return HTML representation.
+   */
+  public String diff_prettyHtml(LinkedList<Diff> diffs) {
+    StringBuilder html = new StringBuilder();
+    for (Diff aDiff : diffs) {
+      String text = aDiff.text.replace("&", "&amp;").replace("<", "&lt;")
+          .replace(">", "&gt;").replace("\n", "&para;<br>");
+      switch (aDiff.operation) {
+      case INSERT:
+        html.append("<ins style=\"background:#e6ffe6;\">").append(text)
+            .append("</ins>");
+        break;
+      case DELETE:
+        html.append("<del style=\"background:#ffe6e6;\">").append(text)
+            .append("</del>");
+        break;
+      case EQUAL:
+        html.append("<span>").append(text).append("</span>");
+        break;
+      }
+    }
+    return html.toString();
+  }
+
+  /**
+   * Compute and return the source text (all equalities and deletions).
+   * @param diffs LinkedList of Diff objects.
+   * @return Source text.
+   */
+  public String diff_text1(LinkedList<Diff> diffs) {
+    StringBuilder text = new StringBuilder();
+    for (Diff aDiff : diffs) {
+      if (aDiff.operation != Operation.INSERT) {
+        text.append(aDiff.text);
+      }
+    }
+    return text.toString();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1432:1438, 1447:1453.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.